### ISSUE TITLE
 "double free or corruption" error

### ISSUE THREAD

**Issue by mauriziocingi** [OpenWaterAnalytics/EPANET#42]
In case of [TIMES] Statistic not equal NONE (see page 174 of Epanet2 users manual) ENclose() function tries to close TmpOutFile again giving a "double free or corruption" error.

You can reproduce error modifing example-networks/Net1.inp at line 124 
from ```Statistic None``` 
to ```Statistic AVERAGED``` and running epanet.

**Solved** adding 
```TmpOutFile=NULL;``` at line 506 in src/output.c

(included in pull request #39 )

**Comment by mauriziocingi** [comment]
It is a bug related to gcc/Linux. 
gcc/mingw32 does not have any problem with fclose() on already closed file.

- epanet2.dll (original from EPA) no problem
- epanet2.dll (from OpenWaterAnalytics compiled with gcc/mingw32) no problem
- epanet2.so (from OpenWaterAnalytics compiled with gcc/Linux) **double free or corruption**

**Comment by mauriziocingi** [comment]
output from  (from OpenWaterAnalytics compiled with gcc/Linux)  where Net1.inp is example-networks/Net1.inp modified at line 124 from ```Statistic None``` to ```Statistic AVERAGED```


```
maurizio@VirtualBox:~/EPANET/build/Linux$ /usr/local/bin/epanet2 Net1.inp temp.txt temp.bin 

EPANET Version 2.1.0

Retrieving network data...
Computing hydraulics at hour 0:00:00
Computing hydraulics at hour 1:00:00
  [omissis]
Computing hydraulics at hour 24:00:00
Computing water quality at hour 0:00:00
  [omissis]
Computing water quality at hour 24:00:00
*** Error in `/usr/local/bin/epanet2': double free or corruption (!prev): 0x092e4218 ***
Aborted (core dumped)

```

**Comment by samhatchett** [comment]
not sure i totally get this. this change https://github.com/OpenWaterAnalytics/EPANET/pull/39/files#diff-1e3dd7877c487d2b3b9f07d9e14e6637R506 sets the file handle to NULL after it is fclose()'d, right?

**Comment by mauriziocingi** [comment]
Yes, I'll explain it better....
While using statistics option output.c uses file TmpOutFile fclose () it.
The ENclose () if TmpOutFile!=NULL fclose () it again the sets TmpOutFile
to NULL.
This is not a problem with Windows but with Linux risse error "double
free....".

I see two possibile solutions:

- output.c dose not fclose () TmpOutFile
- output.c sets TmpOutFile to NULL after fclose () as flag for not close
again (rest of our code does in that way)

The second one sound better
Il 26/Nov/2015 01:40, "Sam Hatchett" <notifications@github.com> ha scritto:

**Comment by samhatchett** [comment]
Got it. Thanks for the explanation. I was assuming the crash was at *that* fclose when in fact it was later down the line. Good fix.

**Comment by samhatchett** [comment]
also can report that this was not a crashing issue on Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)

**Comment by samhatchett** [comment]
sounds like this is fixed and queued for 2.1, so i'm closing.
